components/validator/guardian.go

```diff
@@ -91,16 +93,61 @@ func (g *Guardian) Stop() error {
 }
 
 func (g *Guardian) ValidateL2Output(ctx context.Context, outputRoot eth.Bytes32, l2BlockNumber uint64) (bool, error) {
+	g.log.Info("validating output...", "blockNumber", l2BlockNumber, "outputRoot", outputRoot)
 	localOutputRoot, err := g.outputRootAtBlock(ctx, l2BlockNumber)
 	if err != nil {
-		return false, fmt.Errorf("failed to get outputRootAtBlock: %w", err)
+		return false, fmt.Errorf("failed to get output root at block %d: %w", l2BlockNumber, err)
 	}
 	isValid := bytes.Equal(outputRoot[:], localOutputRoot[:])
 	return isValid, nil
 }
 
 func (g *Guardian) ConfirmTransaction(ctx context.Context, transactionId *big.Int) (*types.Transaction, error) {
-	txOpts := utils.NewSimpleTxOpts(ctx, g.cfg.TxManager.From(), g.cfg.TxManager.Signer)
+	cCtx, cCancel := context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	executionTx, err := g.securityCouncilContract.Transactions(&bind.CallOpts{Context: cCtx}, transactionId)
+	cCancel()
+	if err != nil {
+		return nil, err
+	}
+	if executionTx.Executed {
+		log.Warn("transaction %d already executed.", transactionId)
+		return nil, nil
+	}
+
+	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	executionGas, err := g.cfg.L1Client.EstimateGas(cCtx, ethereum.CallMsg{
+		From:  g.cfg.SecurityCouncilAddr,
+		To:    &executionTx.Destination,
+		Value: executionTx.Value,
+		Data:  executionTx.Data,
+	})
+	cCancel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to estimate confirmation tx gas: %w", err)
+	}
+
+	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	txOpts := utils.NewSimpleTxOpts(cCtx, g.cfg.TxManager.From(), g.cfg.TxManager.Signer)
+	confirmTx, err := g.securityCouncilContract.ConfirmTransaction(txOpts, transactionId)
+	cCancel()
+	if err != nil {
+		return nil, err
+	}
+
+	cCtx, cCancel = context.WithTimeout(ctx, g.cfg.NetworkTimeout)
+	confirmationGas, err := g.cfg.L1Client.EstimateGas(cCtx, ethereum.CallMsg{
+		From:  txOpts.From,
+		To:    confirmTx.To(),
+		Data:  confirmTx.Data(),
+		Value: confirmTx.Value(),
+	})
+	cCancel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to estimate confirmation tx gas: %w", err)
+	}
+
+	txOpts.Context = ctx
+	txOpts.GasLimit = (confirmationGas * GasLimitMultiplier / GasLimitDenominator) + executionGas
```

Comment:
I think it would be clearer if we declared just one GasLimitMultiplier as a float type.
```suggestion
    const (
        GasLimitMultiplier float32 = 1.5
    )
	txOpts.GasLimit = uint64(float32(confirmationGas) * GasLimitMultiplier) + executionGas
```